sys/include/fido2/ctap/transport/hid/ctap_transport_hid.h

```diff
@@ -0,0 +1,261 @@
+ */
+typedef struct {
+    uint8_t seq;                                    /**< packet sequence number */
+    uint8_t payload[CTAP_HID_CONT_PAYLOAD_SIZE];    /**< packet payload */
+} ctap_hid_cont_pkt_t;
+
+/**
+ * @brief CTAP_HID packet struct
+ *
+ */
+typedef struct {
+    uint32_t cid;                   /**< channel identifier */
+    union {
+        ctap_hid_init_pkt_t init;   /**< initialization packet */
+        ctap_hid_cont_pkt_t cont;   /**< continuation packet */
+    };
+} ctap_hid_pkt_t;
+
+/**
+ * @brief CTAP_HID initialization response struct
+ *
+ * CTAP specification (version 20190130) 8.1.9.1.3
+ */
+typedef struct __attribute__((packed)){
+    uint8_t nonce[CTAP_HID_INIT_NONCE_SIZE];    /**< nonce */
+    uint32_t cid;                               /**< channel identifier */
+    uint8_t protocol_version;                   /**< CTAP_HID protocol version */
+    uint8_t version_major;                      /**< major device version */
+    uint8_t version_minor;                      /**< minor device version */
+    uint8_t build_version;                      /**< build device version */
+    uint8_t capabilities;                       /**< capabilities flags */
+} ctap_hid_init_resp_t;
+
+/**
+ * @brief CTAP_HID channel identifier struct
+ *
+ * Used to keep state information about logical channels
+ */
+typedef struct {
+    bool taken;                             /**< is cid taken? */
+    uint32_t cid;                           /**< channel identifier */
+    uint64_t last_used;                     /**< timestamp of last usage */
+} ctap_hid_cid_t;
+
+/**
+ * @brief CTAP_HID buffer struct
+ *
+ */
+typedef struct {
+    uint32_t cid;                           /**< channel identifier */
+    uint8_t cmd;                            /**< CTAP_HID command */
+    uint8_t buffer[CTAP_HID_BUFFER_SIZE];   /**< data buffer */
+    uint16_t offset;                        /**< current offset into data buffer */
+    int16_t seq;                            /**< current sequence number */
+    uint16_t bcnt;                          /**< expected amount of bytes to be received */
+    uint8_t err;                            /**< error type if error  */
+    bool is_locked;                         /**< buffer is locked by transaction */
+    bool should_cancel;                     /**< current transaction should be cancelled */
+} ctap_hid_buffer_t;
+
+/**
+ * @brief Initialize USB, create needed threads
+ *
+ */
+void fido2_ctap_transport_hid_create(void);
+
+/**
+ * @brief Handle CTAP_HID packet
+ *
+ * @param[in] pkt_raw   CTAP_HID packet
+ * @param[in] size      size of packet
+ *
+ */
+void fido2_ctap_transport_hid_handle_packet(void *pkt_raw, int size);
+
+/**
+ * @brief Send keepalive packet
+ *
+ * @param[in] status    CTAP_HID status code
+ *
+ */
+void fido2_ctap_transport_hid_send_keepalive(uint8_t status);
+
+/**
+ * @brief Check logical channels for timeouts
+ *
+ */
+void fido2_ctap_transport_hid_check_timeouts(void);
```

Comment:
It is unclear to me why and how this function should be used. Should it be an internal function?

Reply:
This function is used to prevent one channel from locking the authenticator. E.g. if a device starts a transaction that does not fit in one packet and sends a CTAPHID initialization packet but not continuation packet the authenticator will keep waiting. `fido2_ctap_transport_hid_check_timeouts` will prevent this by cancelling a transaction if it takes too long

Reply:
1. So should this rather be an internal function?
2. How about extending the documentation with your explanation above?

Reply:
I will extend the documentation.

The function can not be internal since it needs to be called regularly. It needs to be called even if `fido2_ctap_transport_read_timeout` didn't receive anything from the USB layer.